File: pkg/protocol/sign.go
package protocol

import (
	"crypto"
	"crypto/ecdh"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"math/big"
)

// ErrSignature is returned when signature verification fails.
var ErrSignature = errors.New("invalid signature")

// p256PointLen is the length of an uncompressed P-256 point: a 0x04 tag plus
// two 32-byte coordinates. This is what PSA's psa_export_public_key emits for
// a P-256 key, so a constrained device can put its exported bytes straight on
// the wire.
const p256PointLen = 65

// p256SigLen is the length of a raw (r || s) P-256 signature, each value
// left-padded to the 32-byte field size.
//
// Deliberately NOT ASN.1/DER: PSA's psa_sign_hash emits raw r||s, and asking a
// microcontroller to wrap that in DER — or to parse DER coming back — costs
// code for no benefit. The Go side converts instead.
const p256SigLen = 64

// Sign canonicalises v, signs the bytes with priv, and returns a SignedEnvelope.
// keyID is an opaque label callers use to identify the signer (e.g. "device",
// "server-2026-04").
//
// This is the Ed25519 entry point, kept for every caller and agent that has
// always used it. Use SignWithSuite to sign under a specific suite.
func Sign(v any, priv ed25519.PrivateKey, keyID string) (*SignedEnvelope, error) {
	return SignWithSuite(v, priv, keyID, SuiteEd25519X25519)
}

// SignWithSuite canonicalises v and signs it under the given suite. key must
// match the suite: an ed25519.PrivateKey for SuiteEd25519X25519, an
// *ecdsa.PrivateKey on P-256 for SuiteP256.
func SignWithSuite(v any, key crypto.Signer, keyID string, suite Suite) (*SignedEnvelope, error) {
	canon, err := Canonicalize(v)
	if err != nil {
		return nil, err
	}
	sig, err := signCanonical(canon, key, suite)
	if err != nil {
		return nil, err
	}
	return &SignedEnvelope{
		ProtocolVersion: Version,
		KeyID:           keyID,
		Algorithm:       suite.SignAlg(),
		Payload:         base64.StdEncoding.EncodeToString(canon),
		Signature:       base64.StdEncoding.EncodeToString(sig),
	}, nil
}

func signCanonical(canon []byte, key crypto.Signer, suite Suite) ([]byte, error) {
	switch suite {
	case SuiteP256:
		priv, ok := key.(*ecdsa.PrivateKey)
		if !ok {
			return nil, fmt.Errorf("suite %s needs an *ecdsa.PrivateKey, got %T", suite, key)
		}
		if priv.Curve != elliptic.P256() {
			return nil, fmt.Errorf("suite %s needs a P-256 key", suite)
		}
		digest := sha256.Sum256(canon)
		r, s, err := ecdsa.Sign(rand.Reader, priv, digest[:])
		if err != nil {
			return nil, err
		}
		return encodeP256Signature(r, s), nil
	default:
		priv, ok := key.(ed25519.PrivateKey)
		if !ok {
			return nil, fmt.Errorf("suite %s needs an ed25519.PrivateKey, got %T", suite, key)
		}
		return ed25519.Sign(priv, canon), nil
	}
}

// Verify checks the signature on env using pub and, on success, returns the
// canonical payload bytes for the caller to JSON-decode.
//
// pub is an ed25519.PublicKey or an *ecdsa.PublicKey, matching env.Algorithm.
// The parameter is crypto.PublicKey (an alias for any) rather than a concrete
// type so that existing Ed25519 callers compile unchanged.
func Verify(env *SignedEnvelope, pub crypto.PublicKey) ([]byte, error) {
	if env == nil {
		return nil, errors.New("nil envelope")
	}
	payload, err := base64.StdEncoding.DecodeString(env.Payload)
	if err != nil {
		return nil, fmt.Errorf("decode payload: %w", err)
	}
	sig, err := base64.StdEncoding.DecodeString(env.Signature)
	if err != nil {
		return nil, fmt.Errorf("decode signature: %w", err)
	}

	switch env.Algorithm {
	case AlgEd25519:
		key, ok := pub.(ed25519.PublicKey)
		if !ok {
			return nil, fmt.Errorf("alg %s needs an ed25519.PublicKey, got %T", env.Algorithm, pub)
		}
		if !ed25519.Verify(key, payload, sig) {
			return nil, ErrSignature
		}
	case AlgECDSAP256:
		key, ok := pub.(*ecdsa.PublicKey)
		if !ok {
			return nil, fmt.Errorf("alg %s needs an *ecdsa.PublicKey, got %T", env.Algorithm, pub)
		}
		r, s, err := decodeP256Signature(sig)
		if err != nil {
			return nil, err
		}
		digest := sha256.Sum256(payload)
		if !ecdsa.Verify(key, digest[:], r, s) {
			return nil, ErrSignature
		}
	default:
		return nil, fmt.Errorf("unsupported alg %q", env.Algorithm)
	}
	return payload, nil
}

// DecodePayloadUnverified base64-decodes the payload from env without
// checking the signature. Used for BLE TOFU mode when no server pubkey is
// configured. Callers should log a warning before calling this.
func DecodePayloadUnverified(env *SignedEnvelope) ([]byte, error) {
	if env == nil {
		return nil, errors.New("nil envelope")
	}
	payload, err := base64.StdEncoding.DecodeString(env.Payload)
	if err != nil {
		return nil, fmt.Errorf("decode payload: %w", err)
	}
	return payload, nil
}

// NewNonce returns a base64-encoded 16-byte random nonce suitable for use in
// EnrollRequest.Nonce.
func NewNonce() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(b), nil
}

// EncodePublicKey returns the base64 form used on the wire for an Ed25519
// public key.
func EncodePublicKey(pub ed25519.PublicKey) string {
	return base64.StdEncoding.EncodeToString(pub)
}

// EncodePublicKeyForSuite returns the base64 wire form of a public key under
// the given suite: a raw 32-byte Ed25519 key, or an uncompressed P-256 point.
func EncodePublicKeyForSuite(pub crypto.PublicKey, suite Suite) (string, error) {
	switch suite {
	case SuiteP256:
		key, ok := pub.(*ecdsa.PublicKey)
		if !ok {
			return "", fmt.Errorf("suite %s needs an *ecdsa.PublicKey, got %T", suite, pub)
		}
		return base64.StdEncoding.EncodeToString(encodeP256Point(key)), nil
	default:
		key, ok := pub.(ed25519.PublicKey)
		if !ok {
			return "", fmt.Errorf("suite %s needs an ed25519.PublicKey, got %T", suite, pub)
		}
		return base64.StdEncoding.EncodeToString(key), nil
	}
}

// DecodePublicKey parses the wire form of an Ed25519 public key.
func DecodePublicKey(s string) (ed25519.PublicKey, error) {
	b, err := base64.StdEncoding.DecodeString(s)
	if err != nil {
		return nil, err
	}
	if len(b) != ed25519.PublicKeySize {
		return nil, fmt.Errorf("expected %d bytes, got %d", ed25519.PublicKeySize, len(b))
	}
	return ed25519.PublicKey(b), nil
}

// DecodePublicKeyForAlg parses the wire form of a public key, choosing the key
// type from an envelope's algorithm. The result is suitable to pass to Verify.
func DecodePublicKeyForAlg(s, alg string) (crypto.PublicKey, error) {
	switch alg {
	case AlgEd25519, "":
		return DecodePublicKey(s)
	case AlgECDSAP256:
		b, err := base64.StdEncoding.DecodeString(s)
		if err != nil {
			return nil, err
		}
		return decodeP256Point(b)
	default:
		return nil, fmt.Errorf("unsupported alg %q", alg)
	}
}

// encodeP256Point renders a public key as an uncompressed SEC 1 point.
func encodeP256Point(pub *ecdsa.PublicKey) []byte {
	out := make([]byte, p256PointLen)
	out[0] = 4
	pub.X.FillBytes(out[1:33])
	pub.Y.FillBytes(out[33:65])
	return out
}

// decodeP256Point parses an uncompressed SEC 1 point and rejects anything not
// on the curve.
//
// Point validation goes through crypto/ecdh, whose NewPublicKey performs the
// on-curve check; elliptic.IsOnCurve and elliptic.Unmarshal are both
// deprecated. The parsed coordinates are then reused to build the
// ecdsa.PublicKey that ecdsa.Verify needs.
func decodeP256Point(b []byte) (*ecdsa.PublicKey, error) {
	if len(b) != p256PointLen || b[0] != 4 {
		return nil, fmt.Errorf("expected a %d-byte uncompressed P-256 point, got %d bytes",
			p256PointLen, len(b))
	}
	if _, err := ecdh.P256().NewPublicKey(b); err != nil {
		return nil, fmt.Errorf("invalid P-256 point: %w", err)
	}
	return &ecdsa.PublicKey{
		Curve: elliptic.P256(),
		X:     new(big.Int).SetBytes(b[1:33]),
		Y:     new(big.Int).SetBytes(b[33:65]),
	}, nil
}

// encodeP256Signature renders (r, s) as the raw 64-byte form PSA emits.
func encodeP256Signature(r, s *big.Int) []byte {
	out := make([]byte, p256SigLen)
	r.FillBytes(out[:32])
	s.FillBytes(out[32:])
	return out
}

func decodeP256Signature(sig []byte) (r, s *big.Int, err error) {
	if len(sig) != p256SigLen {
		return nil, nil, fmt.Errorf("expected a %d-byte raw P-256 signature, got %d bytes",
			p256SigLen, len(sig))
	}
	return new(big.Int).SetBytes(sig[:32]), new(big.Int).SetBytes(sig[32:]), nil
}

File: clients/rust/src/sign.rs
//! Envelope sign / verify — matches pkg/protocol/sign.go.
//!
//! Signing input = RFC 8785 canonical JSON bytes of the payload struct.
//! The signature covers exactly those bytes (no additional framing).
//!
//! Two algorithms, one per [`Suite`]:
//! - `ed25519` — 32-byte public keys, 64-byte signatures.
//! - `ecdsa-p256-sha256` — uncompressed 65-byte SEC 1 public keys and raw
//!   64-byte `r‖s` signatures (no ASN.1), the forms PSA emits.

use base64::{engine::general_purpose::STANDARD, Engine as _};
use ed25519_dalek::{Signer as _, Verifier as _};
use rand::rngs::OsRng;
use serde::{Deserialize, Serialize};

use crate::canonical;
use crate::suite::{Suite, ALG_ECDSA_P256, ALG_ED25519};

/// Length of an uncompressed P-256 point: 0x04 tag + two 32-byte coordinates.
const P256_POINT_LEN: usize = 65;

/// Wire-format signed envelope (matches protocol.SignedEnvelope in wire.go).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SignedEnvelope {
    pub protocol_version: String,
    pub key_id: String,
    pub alg: String,       // "ed25519" | "ecdsa-p256-sha256"
    pub payload: String,   // base64(canonical JSON)
    pub signature: String, // base64(signature)
}

/// A signing key under one of the supported suites.
#[derive(Clone)]
pub enum PrivateKey {
    Ed25519(ed25519_dalek::SigningKey),
    P256(p256::ecdsa::SigningKey),
}

impl PrivateKey {
    /// Generate a fresh key for `suite`.
    pub fn generate(suite: Suite) -> Self {
        match suite {
            Suite::Ed25519X25519 => PrivateKey::Ed25519(ed25519_dalek::SigningKey::generate(&mut OsRng)),
            Suite::P256 => PrivateKey::P256(p256::ecdsa::SigningKey::random(&mut OsRng)),
        }
    }

    pub fn suite(&self) -> Suite {
        match self {
            PrivateKey::Ed25519(_) => Suite::Ed25519X25519,
            PrivateKey::P256(_) => Suite::P256,
        }
    }

    pub fn public_key(&self) -> PublicKey {
        match self {
            PrivateKey::Ed25519(k) => PublicKey::Ed25519(k.verifying_key()),
            PrivateKey::P256(k) => PublicKey::P256(*k.verifying_key()),
        }
    }

    fn sign_bytes(&self, msg: &[u8]) -> Vec<u8> {
        match self {
            PrivateKey::Ed25519(k) => k.sign(msg).to_bytes().to_vec(),
            PrivateKey::P256(k) => {
                // SHA-256 digest + RFC 6979 nonce; to_bytes() is raw r‖s.
                let sig: p256::ecdsa::Signature = p256::ecdsa::signature::Signer::sign(k, msg);
                sig.to_bytes().to_vec()
            }
        }
    }
}

impl From<ed25519_dalek::SigningKey> for PrivateKey {
    fn from(k: ed25519_dalek::SigningKey) -> Self {
        PrivateKey::Ed25519(k)
    }
}

impl From<p256::ecdsa::SigningKey> for PrivateKey {
    fn from(k: p256::ecdsa::SigningKey) -> Self {
        PrivateKey::P256(k)
    }
}

/// A verifying key under one of the supported suites.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PublicKey {
    Ed25519(ed25519_dalek::VerifyingKey),
    P256(p256::ecdsa::VerifyingKey),
}

impl PublicKey {
    pub fn suite(&self) -> Suite {
        match self {
            PublicKey::Ed25519(_) => Suite::Ed25519X25519,
            PublicKey::P256(_) => Suite::P256,
        }
    }

    /// Raw wire bytes: 32-byte Ed25519 key or 65-byte uncompressed P-256 point.
    pub fn to_bytes(&self) -> Vec<u8> {
        match self {
            PublicKey::Ed25519(k) => k.as_bytes().to_vec(),
            PublicKey::P256(k) => k.to_encoded_point(false).as_bytes().to_vec(),
        }
    }

    /// Parse raw wire bytes, choosing the key type by length.
    pub fn from_bytes(b: &[u8]) -> crate::Result<Self> {
        match b.len() {
            32 => {
                let arr: [u8; 32] = b.try_into().unwrap();
                ed25519_dalek::VerifyingKey::from_bytes(&arr)
                    .map(PublicKey::Ed25519)
                    .map_err(|e| format!("invalid Ed25519 public key: {e}").into())
            }
            P256_POINT_LEN if b[0] == 4 => p256::ecdsa::VerifyingKey::from_sec1_bytes(b)
                .map(PublicKey::P256)
                .map_err(|e| format!("invalid P-256 point: {e}").into()),
            n => Err(format!(
                "public key must be a 32-byte Ed25519 key or a {P256_POINT_LEN}-byte \
                 uncompressed P-256 point, got {n} bytes"
            )
            .into()),
        }
    }
}

/// Sign `value` with `key`, returning a [`SignedEnvelope`] whose `alg` follows
/// the key's suite.
///
/// `key_id` is an opaque label for the signer, e.g. `"device"` or `"server"`.
pub fn sign<T: Serialize>(value: &T, key: &PrivateKey, key_id: &str) -> crate::Result<SignedEnvelope> {
    let canon = canonical::canonicalize(value)?;
    let sig = key.sign_bytes(&canon);
    Ok(SignedEnvelope {
        protocol_version: crate::wire::VERSION.to_string(),
        key_id: key_id.to_string(),
        alg: key.suite().sign_alg().to_string(),
        payload: STANDARD.encode(&canon),
        signature: STANDARD.encode(sig),
    })
}

/// Verify the signature on `env` using `pub_key`.
///
/// The algorithm comes from `env.alg`, and must match the pinned key's type:
/// a server that signs under a different suite than the key the device trusts
/// is rejected rather than silently accepted.
///
/// Returns the canonical payload bytes on success so the caller can
/// `serde_json::from_slice` them into the expected type.
pub fn verify(env: &SignedEnvelope, pub_key: &PublicKey) -> crate::Result<Vec<u8>> {
    let alg_suite = Suite::for_sign_alg(&env.alg)?;
    if alg_suite != pub_key.suite() {
        return Err(format!(
            "envelope is signed with {:?} but the trusted server key is for the {} suite \
             (does the server profile's crypto suite match this agent's --crypto-suite?)",
            env.alg,
            pub_key.suite()
        )
        .into());
    }
    let payload = STANDARD.decode(&env.payload)?;
    let sig_bytes = STANDARD.decode(&env.signature)?;
    match (env.alg.as_str(), pub_key) {
        (ALG_ED25519, PublicKey::Ed25519(k)) => {
            let sig_arr: [u8; 64] = sig_bytes
                .as_slice()
                .try_into()
                .map_err(|_| "signature must be 64 bytes")?;
            let sig = ed25519_dalek::Signature::from_bytes(&sig_arr);
            k.verify(&payload, &sig)
                .map_err(|e| -> crate::Error { format!("invalid signature: {e}").into() })?;
        }
        (ALG_ECDSA_P256, PublicKey::P256(k)) => {
            if sig_bytes.len() != 64 {
                return Err(format!(
                    "expected a 64-byte raw P-256 signature, got {} bytes",
                    sig_bytes.len()
                )
                .into());
            }
            let sig = p256::ecdsa::Signature::from_slice(&sig_bytes)
                .map_err(|e| format!("invalid signature: {e}"))?;
            p256::ecdsa::signature::Verifier::verify(k, &payload, &sig)
                .map_err(|e| -> crate::Error { format!("invalid signature: {e}").into() })?;
        }
        _ => unreachable!("alg/key suite checked above"),
    }
    Ok(payload)
}

/// Decode the payload from `env` **without** verifying the signature.
///
/// Only for TOFU / BLE mode where the server pubkey is not yet known.
/// Logs a warning so the caller is reminded that trust is not established.
pub fn decode_payload_unverified(env: &SignedEnvelope) -> crate::Result<Vec<u8>> {
    log::warn!(
        "bundle signature NOT verified (TOFU mode — no --server-pubkey provided). \
         Use --server-pubkey to pin the server's key after first enrollment."
    );
    Ok(STANDARD.decode(&env.payload)?)
}

/// Generate a 16-byte random nonce and return it base64-encoded (standard).
pub fn new_nonce() -> String {
    let mut b = [0u8; 16];
    use rand::RngCore;
    OsRng.fill_bytes(&mut b);
    STANDARD.encode(b)
}

/// Encode a public key as standard base64 (wire format).
pub fn encode_public_key(key: &PublicKey) -> String {
    STANDARD.encode(key.to_bytes())
}

/// Decode a wire-format base64 public key; the suite is inferred from its
/// length (32 bytes → Ed25519, 65 bytes → uncompressed P-256).
pub fn decode_public_key(s: &str) -> crate::Result<PublicKey> {
    PublicKey::from_bytes(&STANDARD.decode(s.trim())?)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn test_key() -> PrivateKey {
        let seed: [u8; 32] = (1u8..=32).collect::<Vec<_>>().try_into().unwrap();
        ed25519_dalek::SigningKey::from_bytes(&seed).into()
    }

    #[test]
    fn sign_verify_roundtrip() {
        for key in [test_key(), PrivateKey::generate(Suite::P256)] {
            let pub_key = key.public_key();
            let payload = json!({"b": "val", "a": 42});
            let env = sign(&payload, &key, "device").unwrap();
            assert_eq!(env.alg, key.suite().sign_alg());
            let raw = verify(&env, &pub_key).unwrap();
            let v: serde_json::Value = serde_json::from_slice(&raw).unwrap();
            assert_eq!(v["a"], json!(42));
        }
    }

    #[test]
    fn tampered_payload_rejected() {
        for key in [test_key(), PrivateKey::generate(Suite::P256)] {
            let pub_key = key.public_key();
            let mut env = sign(&json!({"x": 1}), &key, "k").unwrap();
            // Replace payload with a different value's canonical bytes
            let tampered = canonical::canonicalize(&json!({"x": 2})).unwrap();
            env.payload = STANDARD.encode(&tampered);
            assert!(verify(&env, &pub_key).is_err());
        }
    }

    #[test]
    fn suite_mismatch_rejected() {
        let p256_key = PrivateKey::generate(Suite::P256);
        let env = sign(&json!({"x": 1}), &p256_key, "k").unwrap();
        let err = verify(&env, &test_key().public_key()).unwrap_err();
        assert!(err.to_string().contains("ecdsa-p256-sha256"), "{err}");
    }

    #[test]
    fn public_key_wire_roundtrip() {
        for key in [test_key(), PrivateKey::generate(Suite::P256)] {
            let pub_key = key.public_key();
            let b64 = encode_public_key(&pub_key);
            assert_eq!(decode_public_key(&b64).unwrap(), pub_key);
        }
        assert_eq!(PrivateKey::generate(Suite::P256).public_key().to_bytes().len(), 65);
    }
}
